README.md
| Develop branch | Releases |
|---|----|
| ROS1 Melodic @ u18.04 [![Build Status](http://build.ros.org/job/Mdev__mrpt_bridge__ubuntu_bionic_amd64/badge/icon)](http://build.ros.org/job/Mdev__mrpt_bridge__ubuntu_bionic_amd64/) | ROS1 Melodic @ u18.04: [![Build Status](http://build.ros.org/job/Mbin_uB64__mrpt_bridge__ubuntu_bionic_amd64__binary/badge/icon)](http://build.ros.org/job/Mbin_uB64__mrpt_bridge__ubuntu_bionic_amd64__binary/) |
|  ROS1 Noetic @ u20.04 [![Build Status](https://build.ros.org/job/Ndev__mrpt_bridge__ubuntu_focal_amd64/badge/icon)](https://build.ros.org/job/Ndev__mrpt_bridge__ubuntu_focal_amd64/) | ROS1 Noetic @ u20.04 [![Build Status](https://build.ros.org/job/Nbin_uF64__mrpt_bridge__ubuntu_focal_amd64__binary/badge/icon)](https://build.ros.org/job/Nbin_uF64__mrpt_bridge__ubuntu_focal_amd64__binary/)  |

mrpt_bridge
===============

C++ libraries to convert between ROS and [MRPT](http://www.mrpt.org/) classes and messages.

* ROS wiki: http://wiki.ros.org/mrpt_navigation
* Compiling instructions: http://wiki.ros.org/mrpt_navigation/Tutorials/Installing
* Usage examples and tutorials: http://wiki.ros.org/mrpt_navigation/Tutorials
* Branches:
  * `master`: Intended for use with latest MRPT versions (>=1.5.5, >=2.0.x).


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mrpt_bridge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2022-05-03)
------------------
* Add License file
* Formatting fixes
* Update README.md
* Contributors: Jose Luis Blanco Claraco

1.0.0 (2022-03-04)
------------------
* update dep to mrpt2
* Merge pull request `#12 <https://github.com/mrpt-ros-pkg/mrpt_bridge/issues/12>`_ from MRo47/ros1
  Fixed build error for ubuntu 20.04, ros-noetic in imu.cpp
* fixed bug (array to vec conversion -> use auto)
* Build with opencv 4
* fix build against mrpt2
* update gitignore; add PUBLIC flags to deps
* Quaternion singularity
  Check for degenerate case when converting quaternion to angles.
* Ensure the real part of quaternions is real
  Fixes: https://github.com/mrpt-ros-pkg/pose_cov_ops/issues/7
* time convert: fix round error in nanoseconds
* Fix build against mrpt-2
* Copyright notices, clang-format.
  Also, avoid some bad practices: C includes vs C++ ones; using namespace
  in headers.
* fix build against mrpt2
  Fixed `#5 <https://github.com/mrpt-ros-pkg/mrpt_bridge/issues/5>`_
* Contributors: Jose Luis Blanco Claraco, Jose Luis Blanco-Claraco, Jose-Luis Blanco, Julian Lopez Velasquez, MRo47, Mathieu Mege, cmorcan

0.1.25 (2018-07-11)
-------------------
* Merge pull request `#3 <https://github.com/mrpt-ros-pkg/mrpt_bridge/issues/3>`_ from tuw-robotics/master
  TPose2D to geometry_msgs::Pose
* Update pose.h
* TPose2D Conversion to geometry_msgs::Pose added
* Merge pull request `#2 <https://github.com/mrpt-ros-pkg/mrpt_bridge/issues/2>`_ from clalancette/fix-deb-stretch
  Fix deb stretch
* Fix build warnings for signed/unsigned comparison.
  The width and height of the OccupancyGrid metadata are both
  uint32, so make the loop iterator type match.
  Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
* Workaround Debian stretch not properly setting up Qt dependencies.
  libpcl-dev really has a dependency on Qt5Widgets, but that
  is not expressed by the Debian Stretch packages.  We workaround
  it here by forcing the dependency in both the package.xml and
  the CMakeLists.txt
  Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
* Contributors: Chris Lalancette, Jose Luis Blanco-Claraco, Markus Bader

0.1.24 (2018-06-14)
-------------------
* fix missing namespace
* Reorganize build farm icons
* Update README.md
* Contributors: Jose Luis Blanco-Claraco

0.1.23 (2018-06-08)
-------------------
* make catkin_lint clean
* ROS build farm badges and links
* Contributors: Jose Luis Blanco-Claraco

0.1.22 (2018-05-18)
-------------------
* Release as an independent repository outside of "mrpt_navigation"

0.1.20 (2018-04-26)
-------------------
* fix warnings
* fix build against mrpt 2.0
* partial fix build w mrpt 2.0
* optimized build (-O3)
* Fix travis (`#94 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/94>`_)
  * add dep stereo_msgs
  * add dep stereo_msgs
  * fix minor warnigngs and errors
* Use docker to run modern ROS distros in travis (`#93 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/93>`_)
* fix build errors against mrpt master. Closes `#92 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/92>`_
* Fix install .h files in mrpt-bridge
  Closes `#91 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/91>`_
* Fix CMake error (Closes `#90 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/90>`_)
* marker messages are also logged as beacons if needed
* marker msgs added
* Merge branch 'master' of github.com:tuw-robotics/mrpt_navigation
* Merge pull request `#83 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/83>`_ from raghavendersahdev/master
  Extending ros_bridge to playback MRPT datasets
* Merge branch 'master' into master
* Work done upto August 25 : added CObservationRange
* CMake finds MRPT >=1.5 in ROS master branch
* Merge branch 'master' into compat-mrpt-1.5
* CMake finds MRPT >=1.9
* Work done upto August 21: added stereo image support to mrpt bridge
* small change moved beacon.h to its correct place
* Work done from August 17 to August 18 : added Tester function test the mrpt-ros bridge and publishing ros message after conversion
* changed test_Bridge file, testing the bridge in a different way
* Work done from August 17 to August 18 : added a tester function of the bridge, error while compiling currently
* avoid Eigen warnings with GCC-7
* Removed unnecessry MRPT_VERSION checks
* Fixes for clang format
* Removed c++11 declarations in cmake
* Adapted CMakeLists to new mrpt
* Work done from August 13 to August 14 : added image mrpt-ros bridge and comments
* Work done upto August 13: added mrpt_bridge conversions for range, GPS, IMU under construction; pending: mrpt_bridge for images and comments
* Merge pull request `#74 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/74>`_ from bergercookie/devel
  Use C++11 in mrpt_bridge, mrpt_localization
* Use C++11 in mrpt_bridge, mrpt_localization
* Merge pull request `#72 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/72>`_ from bergercookie/devel
  Add more conversion methods for CNetworkOfPoses classes
* Comply to clang warnings
* Fix compilation issue with forgotten git-related text
* Correct typo
* Work towards mr-slam
* Implement converion method for CNetworkOfPoses2DInf_NA specialization class
* Add more msgs for use in mrpt_graphslam_2d
* fix build with MRPT 1.5.0
* Fix debug messages: (`#65 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/65>`_)
  * replace printf with log_info
  * read and use debug param
  * dumping to console only once
  Also publish pose even if not updating the filter (sorry for mixing commits)
* Fix debug messages:
  * replace printf with log_info
  * read and use debug param
  * dumping to console only once
  Also publish pose even if not updating the filter (sorry for mixing commits)
* documentation patch in mrpt_bridge beacon header (`#62 <https://github.com/mrpt-ros-pkg/mrpt_navigation/issues/62>`_)

src/pose.cpp
/* +------------------------------------------------------------------------+
   |                     Mobile Robot Programming Toolkit (MRPT)            |
   |                          http://www.mrpt.org/                          |
   |                                                                        |
   | Copyright (c) 2005-2018, Individual contributors, see AUTHORS file     |
   | See: http://www.mrpt.org/Authors - All rights reserved.                |
   | Released under BSD License. See details in http://www.mrpt.org/License |
   +------------------------------------------------------------------------+ */

/*
 * pose.cpp
 *
 *  Created on: Mar 14, 2012
 *      Author: Pablo Iñigo Blasco
 *
 *      To understand better how this is implemented see the references:
 *      - http://www.mrpt.org/2D_3D_Geometry
 */

/**
\mainpage

\b mrpt_bridge is a set of functions to convert between common ROS messages and
MRPT C++ classes.

**/

#include <mrpt/poses/CPose2D.h>
#include <mrpt/poses/CPose3D.h>
#include <mrpt/poses/CPosePDFGaussian.h>
#include <mrpt/poses/CPose3DPDFGaussian.h>
#include <mrpt/poses/CPosePDFGaussianInf.h>
#include <mrpt/poses/CPose3DPDFGaussianInf.h>
#include <mrpt/math/CQuaternion.h>
#include <mrpt/math/lightweight_geom_data.h>
#include <geometry_msgs/PoseWithCovariance.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/Quaternion.h>
#include <tf/tf.h>

#include <mrpt/version.h>
#if MRPT_VERSION < 0x199
#include <mrpt/utils/mrpt_macros.h>
#else
#include <mrpt/core/exceptions.h>
#endif

#include "mrpt_bridge/pose.h"

mrpt::math::CMatrixDouble33& mrpt_bridge::convert(
	const tf::Matrix3x3& _src, mrpt::math::CMatrixDouble33& _des)
{
	for (int r = 0; r < 3; r++)
		for (int c = 0; c < 3; c++) _des(r, c) = _src[r][c];
	return _des;
}

tf::Matrix3x3& mrpt_bridge::convert(
	const mrpt::math::CMatrixDouble33& _src, tf::Matrix3x3& _des)
{
	for (int r = 0; r < 3; r++)
		for (int c = 0; c < 3; c++) _des[r][c] = _src(r, c);
	return _des;
}

tf::Transform& mrpt_bridge::convert(
	const mrpt::poses::CPose3DPDFGaussian& _src, tf::Transform& _des)
{
	return convert(_src.mean, _des);
}

geometry_msgs::PoseWithCovariance& mrpt_bridge::convert(
	const mrpt::poses::CPose3DPDFGaussian& _src,
	geometry_msgs::PoseWithCovariance& _des)
{
	convert(_src.mean, _des.pose);

	// Read REP103: http://ros.org/reps/rep-0103.html#covariance-representation
	// # Row-major representation of the 6x6 covariance matrix
	// # The orientation parameters use a fixed-axis representation.
	// # In order, the parameters are:
	// # (x, y, z, rotation about X axis, rotation about Y axis, rotation about
	// Z axis)
	// float64[36] covariance
	// Old comment: "MRPT uses non-fixed axis for 6x6 covariance: should use a
	// transform Jacobian here!"
	//           JL ==> Nope! non-fixed z-y-x equals fixed x-y-z rotations.

	const unsigned int indxs_map[6] = {0, 1, 2,
									   5, 4, 3};  // X,Y,Z,YAW,PITCH,ROLL

	for (int i = 0; i < 6; i++)
	{
		for (int j = 0; j < 6; j++)
		{
			_des.covariance[indxs_map[i] * 6 + indxs_map[j]] = _src.cov(i, j);
		}
	}
	return _des;
}

geometry_msgs::PoseWithCovariance& mrpt_bridge::convert(
	const mrpt::poses::CPose3DPDFGaussianInf& _src,
	geometry_msgs::PoseWithCovariance& _des)
{
	mrpt::poses::CPose3DPDFGaussian mrpt_gaussian;
	mrpt_gaussian.copyFrom(_src);
	convert(mrpt_gaussian, _des);

	return _des;
}

tf::Transform& mrpt_bridge::convert(
	const mrpt::poses::CPose3D& _src, tf::Transform& _des)
{
	tf::Vector3 origin(_src[0], _src[1], _src[2]);
	mrpt::math::CMatrixDouble33 R;
	_src.getRotationMatrix(R);
	tf::Matrix3x3 basis;
	_des.setBasis(convert(R, basis));
	_des.setOrigin(origin);

	return _des;
}
mrpt::poses::CPose3D& mrpt_bridge::convert(
	const tf::Transform& _src, mrpt::poses::CPose3D& _des)
{
	const tf::Vector3& t = _src.getOrigin();
	_des.x() = t[0], _des.y() = t[1], _des.z() = t[2];
	const tf::Matrix3x3& basis = _src.getBasis();
	mrpt::math::CMatrixDouble33 R;
	convert(basis, R);
	_des.setRotationMatrix(R);

	return _des;
}

geometry_msgs::Pose& mrpt_bridge::convert(
	const mrpt::poses::CPose3D& _src, geometry_msgs::Pose& _des)
{
	_des.position.x = _src[0];
	_des.position.y = _src[1];
	_des.position.z = _src[2];

	mrpt::math::CQuaternionDouble q;
	_src.getAsQuaternion(q);

	_des.orientation.x = q.x();
	_des.orientation.y = q.y();
	_des.orientation.z = q.z();
	_des.orientation.w = q.r();

	return _des;
}

/** Convert: MRPT's CPose2D (x,y,yaw) -> ROS's Pose */
geometry_msgs::Pose& mrpt_bridge::convert(
	const mrpt::poses::CPose2D& _src, geometry_msgs::Pose& _des)
{
	_des.position.x = _src.x();
	_des.position.y = _src.y();
	_des.position.z = 0;

	const double yaw = _src.phi();
	if (std::abs(yaw) < 1e-10)
	{
		_des.orientation.x = 0.;
		_des.orientation.y = 0.;
		_des.orientation.z = .5 * yaw;
		_des.orientation.w = 1.;
	}
	else
	{
		const double s = ::sin(yaw * .5);
		const double c = ::cos(yaw * .5);
		_des.orientation.x = 0.;
		_des.orientation.y = 0.;
		_des.orientation.z = s;
		_des.orientation.w = c;
	}

	return _des;
}

/** Convert: MRPT's TPose2D (x,y,yaw) -> ROS's Pose */
geometry_msgs::Pose& mrpt_bridge::convert(
	const mrpt::math::TPose2D& _src, geometry_msgs::Pose& _des)
{
	_des.position.x = _src.x;
	_des.position.y = _src.y;
	_des.position.z = 0;

	const double yaw = _src.phi;
	if (std::abs(yaw) < 1e-10)
	{
		_des.orientation.x = 0.;
		_des.orientation.y = 0.;
		_des.orientation.z = .5 * yaw;
		_des.orientation.w = 1.;
	}
	else
	{
		const double s = ::sin(yaw * .5);
		const double c = ::cos(yaw * .5);
		_des.orientation.x = 0.;
		_des.orientation.y = 0.;
		_des.orientation.z = s;
		_des.orientation.w = c;
	}

	return _des;
}

mrpt::poses::CPose2D& mrpt_bridge::convert(
	const geometry_msgs::Pose& _src, mrpt::poses::CPose2D& _des)
{
	_des.x(_src.position.x);
	_des.y(_src.position.y);

	mrpt::math::CQuaternionDouble quat;
	convert(_src.orientation, quat);

	double roll, pitch, yaw;
	quat.rpy(roll, pitch, yaw);

	_des.phi(yaw);

	return _des;
}

geometry_msgs::PoseWithCovariance& mrpt_bridge::convert(
	const mrpt::poses::CPosePDFGaussian& _src,
	geometry_msgs::PoseWithCovariance& _des)
{
	convert(_src.mean, _des.pose);

	// Read REP103: http://ros.org/reps/rep-0103.html#covariance-representation
	// Old comment: "MRPT uses non-fixed axis for 6x6 covariance: should use a
	// transform Jacobian here!"
	//           JL ==> Nope! non-fixed z-y-x equals fixed x-y-z rotations.

	// geometry_msgs/PoseWithCovariance msg stores the covariance matrix in
	// row-major representation
	// Indexes are :
	// [ 0   1   2   3   4   5  ]
	// [ 6   7   8   9   10  11 ]
	// [ 12  13  14  15  16  17 ]
	// [ 18  19  20  21  22  23 ]
	// [ 24  25  26  27  28  29 ]
	// [ 30  31  32  33  34  35 ]

	_des.covariance[0] = _src.cov(0, 0);
	_des.covariance[1] = _src.cov(0, 1);
	_des.covariance[5] = _src.cov(0, 2);
	_des.covariance[6] = _src.cov(1, 0);
	_des.covariance[7] = _src.cov(1, 1);
	_des.covariance[11] = _src.cov(1, 2);
	_des.covariance[30] = _src.cov(2, 0);
	_des.covariance[31] = _src.cov(2, 1);
	_des.covariance[35] = _src.cov(2, 2);

	return _des;
}

geometry_msgs::PoseWithCovariance& mrpt_bridge::convert(
	const mrpt::poses::CPosePDFGaussianInf& _src,
	geometry_msgs::PoseWithCovariance& _des)
{
	mrpt::poses::CPosePDFGaussian mrpt_gaussian;
	mrpt_gaussian.copyFrom(_src);

	convert(mrpt_gaussian, _des);
	return _des;
}

mrpt::poses::CPose3DPDFGaussian& mrpt_bridge::convert(
	const geometry_msgs::PoseWithCovariance& _src,
	mrpt::poses::CPose3DPDFGaussian& _des)
{
	convert(_src.pose, _des.mean);

	const unsigned int indxs_map[6] = {0, 1, 2, 5, 4, 3};

	for (int i = 0; i < 6; i++)
	{
		for (int j = 0; j < 6; j++)
		{
			_des.cov(i, j) = _src.covariance[indxs_map[i] * 6 + indxs_map[j]];
		}
	}

	return _des;
}

mrpt::poses::CPose3DPDFGaussianInf& mrpt_bridge::convert(
	const geometry_msgs::PoseWithCovariance& _src,
	mrpt::poses::CPose3DPDFGaussianInf& _des)
{
	mrpt::poses::CPose3DPDFGaussian mrpt_gaussian;
	convert(
		_src, mrpt_gaussian);  // Intermediate transform => CPose3DPDFGaussian
	_des.copyFrom(mrpt_gaussian);

	return _des;
}

mrpt::poses::CPosePDFGaussian& mrpt_bridge::convert(
	const geometry_msgs::PoseWithCovariance& _src,
	mrpt::poses::CPosePDFGaussian& _des)
{
	convert(_src.pose, _des.mean);

	_des.cov(0, 0) = _src.covariance[0];
	_des.cov(0, 1) = _src.covariance[1];
	_des.cov(0, 2) = _src.covariance[5];
	_des.cov(1, 0) = _src.covariance[0 + 6];
	_des.cov(1, 1) = _src.covariance[1 + 6];
	_des.cov(1, 2) = _src.covariance[5 + 6];
	_des.cov(2, 0) = _src.covariance[0 + 30];
	_des.cov(2, 1) = _src.covariance[1 + 30];
	_des.cov(2, 2) = _src.covariance[5 + 30];

	return _des;
}

mrpt::poses::CPosePDFGaussianInf& mrpt_bridge::convert(
	const geometry_msgs::PoseWithCovariance& _src,
	mrpt::poses::CPosePDFGaussianInf& _des)
{
	mrpt::poses::CPosePDFGaussian mrpt_gaussian;
	convert(_src, mrpt_gaussian);  // intermediate transform: PoseWithCovariance
	// => CPosePDFGaussian
	_des.copyFrom(mrpt_gaussian);

	return _des;
}

mrpt::poses::CQuaternionDouble& mrpt_bridge::convert(
	const geometry_msgs::Quaternion& _src, mrpt::poses::CQuaternionDouble& _des)
{
	_des.x(_src.x);
	_des.y(_src.y);
	_des.z(_src.z);
	_des.r(_src.w);
	// Ensure the real part of the quaternion is >=0.
	// It seems ROS does not always ensure it.
	if (_des.r() < 0)
	{
		_des.r(-_des.r());
		_des.x(-_des.x());
		_des.y(-_des.y());
		_des.z(-_des.z());
	}
	return _des;
}

geometry_msgs::Quaternion& mrpt_bridge::convert(
	const mrpt::poses::CQuaternionDouble& _src, geometry_msgs::Quaternion& _des)
{
	_des.x = _src.x();
	_des.y = _src.y();
	_des.z = _src.z();
	_des.w = _src.r();

	return _des;
}

mrpt::poses::CPose3D& mrpt_bridge::convert(
	const geometry_msgs::Pose& _src, mrpt::poses::CPose3D& _des)
{
	mrpt::math::CQuaternionDouble q(
		_src.orientation.w, _src.orientation.x, _src.orientation.y,
		_src.orientation.z);
	// Ensure the real part of the quaternion is >=0.
	// It seems ROS does not always ensure it.
	if (q.r() < 0)
	{
		q.r(-q.r());
		q.x(-q.x());
		q.y(-q.y());
		q.z(-q.z());
	}

	_des = mrpt::poses::CPose3D(
		q, _src.position.x, _src.position.y, _src.position.z);

	return _des;
}

tf::Transform& mrpt_bridge::convert(
	const mrpt::math::TPose3D& _src, tf::Transform& _des)
{
	return mrpt_bridge::convert(mrpt::poses::CPose3D(_src), _des);
}

tf::Transform& mrpt_bridge::convert(
	const mrpt::poses::CPose2D& _src, tf::Transform& _des)
{
	return mrpt_bridge::convert(mrpt::poses::CPose3D(_src), _des);
}
tf::Transform& mrpt_bridge::convert(
	const mrpt::math::TPose2D& _src, tf::Transform& _des)
{
	return mrpt_bridge::convert(
		mrpt::poses::CPose3D(mrpt::math::TPose3D(_src)), _des);
}
